subprojects/core/src/main/java/org/gradle/internal/exceptions/ContextAwareException.java

```diff
@@ -45,12 +46,13 @@ public void node(Throwable node) {
         return causes;
     }
 
-    public void accept(ExceptionContextVisitor contextVisitor) {
+    public void accept(ExceptionFormattingVisitor contextVisitor) {
```

Comment:
I am skeptical of the changes made here to accept a specific subclass of the visitor interface. This breaks the decoupling that visitors try to achieve. 

The `printSuppressedDuplicateCountIfNecessary` should either be part of the visitor interface or not and since it is specific to the implementation of the visitor it makes me think this should be organized a bit differently.

Looking at the code that actually _calls_ this accept method, it seems `accept` is only really called in `BuildExceptionReporter#constructFailureDetails`. 

We should be able to call `printSuppressedDuplicateCountIfNecessary` from that method and then we don't need to update these `accept` methods and the `ContextAwareExceptionTest` (among others) to reference `ExceptionFormattingVisitor`. Further, that would allow us to keep `ExceptionFormattingVisitor` private instead of making it public.

Reply:
> I am skeptical of the changes made here to accept a specific subclass of the visitor interface.

It already did this though, as `ExceptionContextVisitor` is a specific subclass of `TreeVisitor`.  But I get your broader point.

I see a few possible solutions:

1. Leave as is.
2. Have the `BuildExceptionReporter` call `printSuppressedDuplicateCountIfNecessary` after calling `accept`, and revert the change to `accept` to have it accept a `ExceptionContextVisitor` as before.
3. As the `ExceptionFormattingVisitor is the only concrete subtype of `ExceptionContextVisitor`, perhaps a better solution would be to merge these two types, and add an `uponFinished` callback to `ExceptionContextVisitor` directly, which would be implemented by the code in `printSuppressedDuplicateCountIfNecessary`, and have `accept` always call it after visiting is complete.

Either 2 or 3 would allow for making `ExceptionFormattingVisitor` private once again.  2 seems like it would require a check-and-cast in `ContextAwareException#accept` if we revert the signature of that method.  I think I like 3 better anyway, as it keeps the iteration pattern the same for all implementations (which can utilize a default no-op `uponFinished` if uninterested).

Going to try that.

Reply:
3 could work but IMO 2 is simpler. Why do you think it would require a cast?

Reply:
If we make ExceptionFormattingVisitor private again, it can't be in the signature of a method in `ContextAwareException`.  Come to think of it, then we couldn't cast it in the body of that method either.

Reply:
Why does it need to be in the signature of a method in `ContextAwareException`? 

I am proposing we do (in `BuildExceptionReporter#constructFailureDetails`):

```
        if (failure instanceof ContextAwareException) {
            ExceptionFormattingVisitor visitor = new ExceptionFormattingVisitor(details)
            ((ContextAwareException) failure).accept(visitor);
            visitor.printSuppressedDuplicateCountIfNecessary()
        } else {
            details.appendDetails();
        }
```

Reply:
That could work, I suppose.  I had assumed that we wouldn't want to split the work of formatting between visiting (`accept`) and another method though - especially considering how the `> There were additional failures...` message is logically and visibly a part of the message that comes out of `accept`.

I'd rather have `accept` perform the complete process of message formatting, which is the complete task of "visiting" that this visitor is meant to accomplish.  Slightly more code changes, but I think it's conceptually simpler.

Reply:
In that case 3 would also work